src/rgw/rgw_rados.cc

```diff
@@ -6306,77 +6306,460 @@ int RGWRados::Object::Read::read(int64_t ofs, int64_t end, bufferlist& bl, optio
+  lock.lock();
+  map<off_t, AioCompletion *>::iterator iter = completion_map.find(ofs);
+  if (iter != completion_map.end()) {
+    AioCompletion *c = iter->second;
+    c->release();
+    completion_map.erase(ofs);
+    io_map.erase(ofs);
+  }
+  lock.unlock();
+
+  // we don't drop a reference here -- e.g., not calling d->put(), because we still
+  // need IoCtx to live, as io callback may still be called
+  //
+}
+
+void get_obj_data::cancel_all_io() {
+  ldout(cct, 20) << "get_obj_data::cancel_all_io()" << dendl;
+  std::lock_guard<std::mutex> l(lock);
+  for (map<off_t, librados::AioCompletion *>::iterator iter = completion_map.begin();
+      iter != completion_map.end(); ++iter) {
+    librados::AioCompletion  *c = iter->second;
+    c->release();
+  }
+}
+
+int get_obj_data::get_complete_ios(off_t ofs, list<bufferlist>& bl_list)
+{
+  std::lock_guard<std::mutex> l(lock);
 
-  get_obj_data(RGWRados* store, RGWGetDataCB* cb, rgw::Aio* aio,
-               uint64_t offset, optional_yield yield)
-    : store(store), client_cb(cb), aio(aio), offset(offset), yield(yield) {}
+  map<off_t, get_obj_io>::iterator liter = io_map.begin();
 
-  int flush(rgw::AioResultList&& results) {
-    int r = rgw::check_for_errors(results);
+  if (liter == io_map.end() ||
+      liter->first != ofs) {
+    return 0;
+  }
+
+  map<off_t, librados::AioCompletion *>::iterator aiter;
+  aiter = completion_map.find(ofs);
+  if (aiter == completion_map.end()) {
+    // completion map does not hold this io, it was cancelled
+    return 0;
+  }
+
+  AioCompletion *completion = aiter->second;
+  int r = completion->get_return_value();
+  if (r < 0)
+    return r;
+
+  for (; aiter != completion_map.end(); ++aiter) {
+    completion = aiter->second;
+    if (!completion->is_safe()) {
+      // reached a request that is not yet complete, stop
+        break;
+    }
+
+    r = completion->get_return_value();
     if (r < 0) {
+      set_cancelled(r); // mark it as cancelled, so that we don't continue processing next operations
       return r;
     }
 
-    auto cmp = [](const auto& lhs, const auto& rhs) { return lhs.id < rhs.id; };
-    results.sort(cmp); // merge() requires results to be sorted first
-    completed.merge(results, cmp); // merge results in sorted order
+    total_read += r;
 
-    while (!completed.empty() && completed.front().id == offset) {
-      auto bl = std::move(completed.front().data);
-      completed.pop_front_and_dispose(std::default_delete<rgw::AioResultEntry>{});
+    map<off_t, get_obj_io>::iterator old_liter = liter++;
+    bl_list.push_back(old_liter->second.bl);
+    io_map.erase(old_liter);
+  }
 
-      offset += bl.length();
-      int r = client_cb->handle_data(bl, 0, bl.length());
-      if (r < 0) {
-        return r;
-      }
-    }
-    return 0;
+  return 0;
+}
+
+static void _cache_aio_completion_cb(sigval_t sigval);
+
+std::vector<string> split(const string &s, const char * delim)
+{
+  stringstream ss(s);
```

Comment:
this is a really expensive way to split strings. please consider using the one from `common/split.h` instead, which doesn't require any heap allocations

Reply:
we've informed regarding this;  aishwarya proposed refactoring after merge;  to be fair, fixing would be easy, but I didn't object to it

Reply:
(because fixing it is now trivial)

Reply:
proposing to replace with `boost::split`

Reply:
this is related to the L2 separate PR, will address it there